go-controller/pkg/ovn/egressip.go

```diff
@@ -246,8 +246,8 @@ func (oc *Controller) deleteNamespacePodsEgressIP(eIP *egressipv1.EgressIP, name
 }
 
 func (oc *Controller) assignEgressIPs(eIP *egressipv1.EgressIP) error {
-	assignments := []egressipv1.EgressIPStatusItem{}
 	oc.eIPAllocatorMutex.Lock()
+	assignments := []egressipv1.EgressIPStatusItem{}
```

Comment:
Sure, it might be that I misinterpreted the data race report.

```
WARNING: DATA RACE
Read at 0x00c0002f9830 by goroutine 26:
  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/egressip/v1.(*EgressIP).DeepCopyInto()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/egressip/v1/zz_generated.deepcopy.go:29 +0x65
  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/egressip/v1.(*EgressIP).DeepCopyObject()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/egressip/v1/zz_generated.deepcopy.go:43 +0xa8
  k8s.io/client-go/testing.(*tracker).Get()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/testing/fixture.go:309 +0x33d
  k8s.io/client-go/testing.ObjectReaction.func1()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/testing/fixture.go:111 +0x45f
  k8s.io/client-go/testing.(*SimpleReactor).React()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/testing/fixture.go:534 +0x66
  k8s.io/client-go/testing.(*Fake).Invokes()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/testing/fake.go:141 +0x30d
  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/egressip/v1/apis/clientset/versioned/typed/egressip/v1/fake.(*FakeEgressIPs).Create()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/crd/egressip/v1/apis/clientset/versioned/typed/egressip/v1/fake/fake_egressip.go:82 +0x237
  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn.glob..func4.5.3.1()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/egressip_local_test.go:278 +0xf51
  github.com/urfave/cli/v2.(*App).RunContext()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/urfave/cli/v2/app.go:315 +0xaa9
  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn.glob..func4.5.3()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/urfave/cli/v2/app.go:215 +0x2c7
  github.com/onsi/ginkgo/internal/leafnodes.(*runner).runSync()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/leafnodes/runner.go:113 +0xdf
  github.com/onsi/ginkgo/internal/leafnodes.(*runner).run()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/leafnodes/runner.go:64 +0x107
  github.com/onsi/ginkgo/internal/leafnodes.(*ItNode).Run()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/leafnodes/it_node.go:26 +0x87
  github.com/onsi/ginkgo/internal/spec.(*Spec).runSample()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/spec/spec.go:215 +0x6e1
  github.com/onsi/ginkgo/internal/spec.(*Spec).Run()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/spec/spec.go:138 +0x167
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpec()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/specrunner/spec_runner.go:200 +0x172
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).runSpecs()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/specrunner/spec_runner.go:170 +0x216
  github.com/onsi/ginkgo/internal/specrunner.(*SpecRunner).Run()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/specrunner/spec_runner.go:66 +0x140
  github.com/onsi/ginkgo/internal/suite.(*Suite).Run()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/internal/suite/suite.go:62 +0x69f
  github.com/onsi/ginkgo.RunSpecsWithCustomReporters()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/ginkgo_dsl.go:226 +0x338
  github.com/onsi/ginkgo.RunSpecsWithDefaultAndCustomReporters()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/ginkgo_dsl.go:214 +0x11f
  github.com/onsi/ginkgo.RunSpecs()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/github.com/onsi/ginkgo/ginkgo_dsl.go:205 +0x1a2
  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn.TestClusterNode()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/ovn_suite_test.go:12 +0xf8
  testing.tRunner()
      /usr/lib/golang/src/testing/testing.go:909 +0x199

Previous write at 0x00c0002f9830 by goroutine 192:
  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn.(*Controller).assignEgressIPs.func1()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/egressip.go:252 +0x77
  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn.(*Controller).assignEgressIPs()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/egressip.go:309 +0xdeb
  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn.(*Controller).addEgressIP()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/egressip.go:38 +0x70d
  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn.(*Controller).WatchEgressIP.func1()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/ovn.go:824 +0x80
  k8s.io/client-go/tools/cache.(*ResourceEventHandlerFuncs).OnAdd()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache/controller.go:218 +0x8c
  k8s.io/client-go/tools/cache.FilteringResourceEventHandler.OnAdd()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache/controller.go:250 +0x7a
  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory.(*Handler).OnAdd()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory/factory.go:57 +0xa6
  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory.(*informer).newFederatedHandler.func1.1()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory/factory.go:263 +0x54
  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory.(*informer).forEachHandler()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory/factory.go:124 +0x1f0
  github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory.(*informer).newFederatedHandler.func1()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory/factory.go:262 +0x8a
  k8s.io/client-go/tools/cache.(*ResourceEventHandlerFuncs).OnAdd()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache/controller.go:218 +0x8c
  k8s.io/client-go/tools/cache.(*processorListener).run.func1()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache/shared_informer.go:744 +0x3ad
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:155 +0x6f
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:156 +0xb3
  k8s.io/apimachinery/pkg/util/wait.JitterUntil()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x10d
  k8s.io/client-go/tools/cache.(*processorListener).run()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:90 +0xa2
  k8s.io/client-go/tools/cache.(*processorListener).run-fm()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache/shared_informer.go:732 +0x41
  k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:73 +0x6a

Goroutine 26 (running) created at:
  testing.(*T).Run()
      /usr/lib/golang/src/testing/testing.go:960 +0x651
  testing.runTests.func1()
      /usr/lib/golang/src/testing/testing.go:1202 +0xa6
  testing.tRunner()
      /usr/lib/golang/src/testing/testing.go:909 +0x199
  testing.runTests()
      /usr/lib/golang/src/testing/testing.go:1200 +0x521
  testing.(*M).Run()
      /usr/lib/golang/src/testing/testing.go:1117 +0x2ff
  main.main()
      _testmain.go:44 +0x223

Goroutine 192 (running) created at:
  k8s.io/apimachinery/pkg/util/wait.(*Group).Start()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71 +0x6f
  k8s.io/client-go/tools/cache.(*sharedProcessor).run.func1()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache/shared_informer.go:590 +0x1b4
  k8s.io/client-go/tools/cache.(*sharedProcessor).run()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache/shared_informer.go:594 +0x3f
  k8s.io/client-go/tools/cache.(*sharedProcessor).run-fm()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache/shared_informer.go:585 +0x4b
  k8s.io/apimachinery/pkg/util/wait.(*Group).StartWithChannel.func1()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:56 +0x45
  k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
      /home/aconstan/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:73 +0x6a
==================

```

Reply:
Yeah, I definitely misinterpreted this race report. I just hit it again with my "fix". I don't understand why it's pointing to line `egressip.go:252` (effectively `eIP.Status.Items = assignments`) and `zz_generated.deepcopy.go:29`

Reply:
OK, I think I _might_ understand it finally. Some tests are creating EgressIP objects, as:

```
_, err := fakeOvn.fakeEgressIPClient.K8sV1().EgressIPs().Create(context.TODO(), &eIP, metav1.CreateOptions{})
```

The hypothesis is that several go-routines are spawned for this when the race detector is set to "on". One of this go routines will be in the process of executing code in `pkg/crd/egressip/v1/apis/clientset/versioned/typed/egressip/v1` while another will be in `egressip.go` and L252 and modifying `eIP` by setting its status. I thus need to modify the tests to have this happen "safely" as I've done for `EgressIPs().Get` and `getEgressIPStatusSafely`, etc.

Reply:
OK, seems fixed now with the additional patch I am working on. Will push soon.

Reply:
so it sounds like this was a problem in the test code and/or something resolved by an additional commit. So is this commit still needed?

Reply:
> I think the problem here isn't really "thread-safety" per se; it's that you're modifying objects that you received from the informer cache

Gahh, of course! Thank you for that pointer, I removed all the locks I added in the tests and switched to modifying the copy and updating the API server with that.

> So is this commit still needed?

I squashed that commit. I still prefer to wrap all operations inside the lock however, so I kept the code.